### PR TITLE
Make PFUser.unlinkWithAuthTypeInBackground() fully async.

### PR DESCRIPTION
**Pull request by nlutsenko**
Implementing a TODO.

### CODE REVIEW

File: Parse/PFUser.m [hunk 1]
```diff
@@ -877,18 +877,19 @@ - (BFTask *)linkWithAuthTypeInBackground:(NSString *)authType authData:(NSDictio
...
-            save = [BFTask taskWithResult:@YES];
-        } else {
-            self.authData[authType] = [NSNull null];
-            dirty = YES;
-            save = [self saveInBackground];
+    return [BFTask taskFromExecutor:[BFExecutor defaultPriorityBackgroundExecutor] withBlock:^id{
+        BFTask *task = nil;
+        @synchronized ([self lock]) {
+            if (!self.authData[authType]) {
+                task = [BFTask taskWithResult:@YES];
```
Comment (richardjrossiii) [inline Parse/PFUser.m]: We can just return from here, why are we putting it into a task variable?

Comment (nlutsenko) [inline Parse/PFUser.m]: Good point. Will change.

**Comment by nlutsenko** [comment]
Updated.

File: Parse/PFUser.m [hunk 2]
```diff
@@ -877,18 +877,16 @@ - (BFTask *)linkWithAuthTypeInBackground:(NSString *)authType authData:(NSDictio
...
+        @synchronized (self.lock) {
+            if (self.authData[authType]) {
+                self.authData[authType] = [NSNull null];
+                dirty = YES;
+            }
+            return [self saveInBackground];
         }
-    }
-    return save;
+        return @YES;
```
Comment (richardjrossiii) [inline Parse/PFUser.m]: I think this logic is wrong. this is unreachable code, and we should only save if we changed the auth data.

**Comment by nlutsenko** [comment]
Right... Good catch. Too much logic for Friday...

**Comment by nlutsenko** [comment]
Updated and ready.

**Pull request merged**